src/topics/resolver/telegram-resolver.ts

```diff
@@ -0,0 +1,160 @@
+
+export class TelegramResolver implements IResolver {
+  private _apiHash: string;
+  private _apiId: string;
+  private _botSession: string | undefined;
+  private _botToken: string;
+  private _client: TelegramClient;
+
+  constructor(
+    apiHash = process.env.TELEGRAM_API_HASH || "",
+    apiId = process.env.TELEGRAM_API_ID || "",
+    botToken = process.env.TELEGRAM_BOT_TOKEN || "",
+    botSession = process.env.TELEGRAM_BOT_SESSION
+  ) {
+    this._apiHash = apiHash;
+    this._apiId = apiId;
+    this._botSession = botSession;
+    this._botToken = botToken;
+  }
+
+  public resolve = async (
+    accounts: FetchedData
+  ): Promise<[FetchedData, FetchedData]> => {
+    const alreadyUpdatedAccounts: FetchedData = {};
+    const alreadyResolvedAccounts: FetchedData = {};
+
+    // extract usernames already resolved
+    const unresolvedAccounts = Object.entries(accounts).filter(
+      ([account, value]) => {
+        if (account.split(":").length === 3) {
+          const id = account.split(":")[2];
+          alreadyResolvedAccounts[resolveAccount("1003", id)] = value;
+          alreadyUpdatedAccounts[account] = value;
+        }
+        return account.split(":").length !== 3;
+      }
+    );
+
+    await this._initTelegramClient();
+    await this._client.connect();
+
+    const resolvedAccountsArrays = await withConcurrency(
+      unresolvedAccounts,
+      this.resolveAccounts,
+      {
+        concurrency: 10,
+        batchSize: 1,
+      }
+    );
+
+    await this._client.disconnect();
+
+    // merge already resolved accounts with the new ones
+    const resolvedAccountsRaw = {
+      ...resolvedAccountsArrays[0],
+      ...alreadyUpdatedAccounts,
+    };
+    const resolvedAccounts = {
+      ...resolvedAccountsArrays[1],
+      ...alreadyResolvedAccounts,
+    };
+
+    return [resolvedAccountsRaw, resolvedAccounts];
+  };
+
+  private _initTelegramClient = async () => {
+    this._client = new TelegramClient(
+      new StringSession(this._botSession), +this._apiId, this._apiHash, {}
+    );
+    this._client.setLogLevel(LogLevel.ERROR);
+    await this._signInIfNeeded();
+  };
+
+  private _signInIfNeeded = async () => {
+    if (this._botSession) {
+      console.log("Session found, no need to authorize bot.")
+    } else {
+      console.log("No session found, authorizing using bot token.");
+      await this._client.start({
+        botAuthToken: this._botToken,
+      });
+      const session = this._client.session.save();
+      console.log("Save session to reuse: " + JSON.stringify(session));
+    }
+  };
+
+  private resolveAccounts = async (
+    accounts: [string, BigNumberish][]
+  ): Promise<[FetchedData, FetchedData]> => {
+    const updatedAccounts: FetchedData = {};
+    const resolvedAccounts: FetchedData = {};
+
+    const prefix = accounts[0][0].split(":")[0];
+
+    // remove 'telegram:' from the accounts
+    const accountWithoutType: [string, BigNumberish][] = accounts.map(
```

Comment:
to follow the logic, I would say that this variable should be named `accountsWithoutType` and to get `usernames` you would then do:
```typescript
const usernames = accountsWithoutType.map((accountWithoutType) => {
      return accountWithoutType[0];
    });
```

The logic is not straightforward when currently looking at the code.

Maybe the best thing is to rename `accountsWithoutType` into `usernamesWithValue` ?
And then you only keep usernames

Reply:
I agree with the rename.

The overall logic of the resolver was copied from the Twitter & Github resolvers. I've done the renaming across all of them so they are aligned/consistent. Maybe we can further refactor them in the future because they seem to have a lot of common logic that can be structured into a BaseResolver.